Lib/shutil.py

```diff
@@ -394,30 +395,33 @@ def _rmtree_unsafe(path, onerror):
 
 # Version using fd-based APIs to protect against races
 def _rmtree_safe_fd(topfd, path, onerror):
-    names = []
     try:
-        names = os.listdir(topfd)
+        with os.scandir(topfd) as scandir_it:
+            entries = list(scandir_it)
     except OSError as err:
         err.filename = path
-        onerror(os.listdir, path, sys.exc_info())
-    for name in names:
-        fullname = os.path.join(path, name)
+        onerror(os.scandir, path, sys.exc_info())
+        return
+    for entry in entries:
+        fullname = os.path.join(path, entry.name)
         try:
-            orig_st = os.stat(name, dir_fd=topfd, follow_symlinks=False)
-            mode = orig_st.st_mode
+            is_dir = entry.is_dir(follow_symlinks=False)
```

Comment:
Can you explain why you are doing this `is_dir` call before re-stat()ing below?

Reply:
If `is_dir` is false, we can avoid calling `stat()`. On Posix `scandir()` usually fills the `is_dir` bit, but `stat()` needs a syscall.

Reply:
But then why is `stat()` needed at all? Isn't `is_dir()` sufficient?

Reply:
Oh, I hadn't noticed the use of `orig_st` below. My bad.